File: Python/20-valid-parentheses/index1.py
class Solution:
  def isValid(self, s: str) -> bool:
    stack = []
    dict = {
      ')': '(',
      ']': '[',
      '}': '{'
    }

    for c in s:
      if c not in dict:
        stack.append(c)
      elif not stack or dict[c] != stack.pop():
        return False

    return not stack

File: JavaScript/20-valid-parentheses/index1.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function(s) {
  const stack = [];
  const dict = {
    ')': '(',
    '}': '{',
    ']': '[',
  };

  for (c of s.split('')) {
    if (!(c in dict)) {
      stack.push(c);
    } else if (!stack.length || dict[c] !== stack.pop()) {
      return false
    }
  }

  return !stack.length;
};
